File: node-service/src/services/logger.js
class Logger {
  constructor(config) {
    this.config = config;
    this.correlationId = null;
  }

  setCorrelationId(correlationId) {
    this.correlationId = correlationId;
  }

  _log(level, message, metadata = {}) {
    const logEntry = {
      timestamp: new Date().toISOString(),
      level,
      message,
      service: 'node-sensor-service',
      ...(this.correlationId && { correlationId: this.correlationId }),
      ...metadata
    };

    const logLevel = this.config.getLogLevel();
    const levels = { error: 0, warn: 1, info: 2, debug: 3 };
    
    if (levels[level] <= levels[logLevel]) {
      console.log(JSON.stringify(logEntry));
    }
  }

  info(message, metadata) {
    this._log('info', message, metadata);
  }

  warn(message, metadata) {
    this._log('warn', message, metadata);
  }

  error(message, metadata) {
    this._log('error', message, metadata);
  }

  debug(message, metadata) {
    this._log('debug', message, metadata);
  }
}

module.exports = Logger;

File: A1/python-service/src/services/logger.py
import json
import logging
from typing import Optional, Dict, Any
from datetime import datetime


class Logger:
    def __init__(self, config):
        self.config = config
        self.correlation_id: Optional[str] = None
        self._setup_logging()

    def _setup_logging(self):
        log_level = self.config.get_log_level().upper()
        logging.basicConfig(
            level=getattr(logging, log_level, logging.INFO),
            format='%(message)s'
        )

    def set_correlation_id(self, correlation_id: str):
        self.correlation_id = correlation_id

    def _log(self, level: str, message: str, metadata: Optional[Dict[str, Any]] = None):
        log_entry = {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'level': level,
            'message': message,
            'service': 'python-sensor-service',
        }

        if self.correlation_id:
            log_entry['correlationId'] = self.correlation_id

        if metadata:
            log_entry.update(metadata)

        # Check if we should log based on configured level
        log_level = self.config.get_log_level().lower()
        levels = {'error': 0, 'warn': 1, 'info': 2, 'debug': 3}
        current_level = levels.get(level, 2)
        configured_level = levels.get(log_level, 2)

        if current_level <= configured_level:
            print(json.dumps(log_entry))

    def info(self, message: str, metadata: Optional[Dict[str, Any]] = None):
        self._log('info', message, metadata)

    def warn(self, message: str, metadata: Optional[Dict[str, Any]] = None):
        self._log('warn', message, metadata)

    def error(self, message: str, metadata: Optional[Dict[str, Any]] = None):
        self._log('error', message, metadata)

    def debug(self, message: str, metadata: Optional[Dict[str, Any]] = None):
        self._log('debug', message, metadata)
